File: py/examples/tinygrad_style_custom_loop_export.py
"""Tinygrad-style custom training loop, then export named entrypoints.

Run from repo root:
  PYTHONPATH=py POLYGRAD_LIB=build/libpolygrad.so python py/examples/tinygrad_style_custom_loop_export.py
"""

import numpy as np

from polygrad import Tensor, nn


class LinearNet:
    def __init__(self):
        self.weight = Tensor([[1.0], [1.0]], requires_grad=True).realize()

    def __call__(self, x):
        return x.dot(self.weight)


model = LinearNet()
opt = nn.optim.Adam([model.weight], lr=0.05)
x = Tensor([[1.0, 2.0]])
y = Tensor([[4.0]])

with Tensor.train():
    losses = []
    for _ in range(6):
        opt.zero_grad()
        loss = (model(x) - y).square().mean()
        loss.backward()
        losses.append(loss.item())
        opt.step()

inp = nn.Input("x", shape=(1, 2))
inst = nn.Model.trace(model, inputs={"x": inp}).export()
out = inst.forward(x=np.array([[1.0, 2.0]], dtype=np.float32))
bundle = inst.save_bundle()

print("forward", out["output"])
print("loss", losses[0], "->", losses[-1])
print("bundle bytes", len(bundle))

File: js/examples/tinygrad_style_custom_loop_export.js
'use strict'

// Run from repo root:
//   POLY_CORE=native node js/examples/tinygrad_style_custom_loop_export.js

const { create } = require('../src')

async function main() {
  const pg = await create({ core: process.env.POLY_CORE || 'auto' })
  const Tensor = pg.Tensor

  class LinearNet {
    constructor() {
      this.weight = new Tensor([[1], [1]])
    }
    call(x) { return x.dot(this.weight) }
  }

  const model = new LinearNet()
  const opt = new pg.nn.optim.Adam([model.weight], { lr: 0.05 })
  const x = new Tensor([[1, 2]])
  const y = new Tensor([[4]])

  for (let i = 0; i < 6; i++) {
    opt.zeroGrad()
    const loss = model.call(x).sub(y).square().mean()
    await loss.backward()
    await opt.step()
  }

  const input = pg.Tensor.empty([1, 2])
  const output = model.call(input)
  const inst = await pg.Instance.fromTensors({
    inputs: { x: input },
    outputs: { output },
    params: { weight: model.weight }
  })
  const out = inst.forward({ x: new Float32Array([1, 2]) })
  console.log('forward', Array.from(out.output))
  console.log('bundle bytes', inst.saveBundle().length)

  await pg.dispose()
}

main().catch(err => {
  console.error(err)
  process.exitCode = 1
})
